crates/agent/src/publications/initialize.rs

```diff
@@ -0,0 +1,100 @@
+use models::Capability;
+use std::future::Future;
+use uuid::Uuid;
+
+/// Initialize a draft prior to build/validation. This may add additional specs to the draft.
+pub trait Initialize {
+    fn initialize(
+        &self,
+        db: &sqlx::PgPool,
+        user_id: Uuid,
+        draft: &mut tables::DraftCatalog,
+    ) -> impl Future<Output = anyhow::Result<()>> + Send;
+}
+
+/// A no-op `Initialize` impl, for when you don't want to expand the draft.
+pub struct NoExpansion;
+impl Initialize for NoExpansion {
+    async fn initialize(
+        &self,
+        _db: &sqlx::PgPool,
+        _user_id: Uuid,
+        _draft: &mut tables::DraftCatalog,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+/// An `Initialize` that expands the draft to touch live specs that read from or write to
+/// any drafted collections. This may optionally filter the specs based on whether the user
+/// has `admin` capability to them.
+pub struct ExpandDraft {
+    /// Whether to filter specs based on the user's capability. If true, then only specs for which
+    /// the user has `admin` capability will be added to the draft.
+    pub filter_user_has_admin: bool,
```

Comment:
I've been debating whether `admin` is too restrictive -- I could see an argument that Alice's update of C should be able to observe / touch-publish a materialization M for which they have read but not admin.

But ultimately I think you're right, and it should be admin, because we don't want to fail Alice's publish of C due to a materialization M that they're unable to change. That's the admin of M's problem, and they'll see a controller error about the issue.

This is a departure from my prior understanding of how spec expansion should work. In the past, I've thought that you _would_ see specs that make use of a read grant you provide for your collection, which you yourself have no grant. Controllers and the eventual consistency they provide have changed the narrative here.

Reply:
Yeah, good call out. Draft expansion is no longer guaranteed to be exhaustive, and we'll rely on controllers to report errors that appear in connected specs that aren't able to be included in the expanded draft.

FWIW, this was already the behavior that's used in prod. The `filter_user_has_admin` variable was added primarily in order to make that behavior more explicit. But I also thought it might be nice to allow disabling that check, so that `ExpandDraft` could at least theoretically be used with a publication having `verify_user_auth: false`. We don't currently have a use case for that, since controller-initiated publications never use spec expansion. But perhaps it could be useful if we ever want to do an exhaustive check to ensure that a given change won't break anything.